### DISCUSSION TITLE
Modifying text with the HTML API in WordPress 6.7

### DISCUSSION THREAD

**Discussion by justintadlock** [WordPress/developer-blog-content#301] [Topic Ideas]
The Tag Processor will let you customize modifiable text in WordPress 6.7. See:

- https://github.com/WordPress/gutenberg/issues/60396
- [Core-61617](https://core.trac.wordpress.org/ticket/61617) Add set_modifiable_text() for plaintext-only modifications.
    - https://github.com/WordPress/wordpress-develop/pull/7007
    - https://github.com/WordPress/wordpress-develop/pull/7150

This will be a good opportunity to follow up the [first article](https://developer.wordpress.org/news/2023/09/28/the-html-api-process-your-tags-not-your-pain/) and the HTML API to show some use cases beyond just modifying attributes. Some quick ideas:

- Inserting text before/after a string.
- Replacing emoticons (e.g., `:)`) with emoji.
- Replacing a "block list" of words with safe words.
- More?

CC: @dmsnell - Pinging for ideas on what you think are solid things to do with this `set_modifiable_text()`?

**Comment by dmsnell** [comment]
> Replacing a "block list" of words with safe words.

This tends to go real poorly in practice - a clbuttic problem in computer science - and tends to negatively impact specific groups of people more than others, and those groups tend to be the ones that are different from the person writing the code.

It might be valuable to be careful about encouraging behaviors which are often unintentionally exclusionary.

> Replacing emoticons (e.g., :)) with emoji.

https://github.com/WordPress/wordpress-develop/pull/6762

`set_modifiable_text()` makes this easier, but the problem is still complicated in Core. This could be a nice example though.

One aspect of `set_modifiable_text()` that is important to acknowledge (and it can be a strength as well as a drawback) is that it inherently respects HTML syntax boundaries. For example, a naïve loop using `get_modifiable_text()` will not detect `:<span>)</span>` as `:)` (which is probably correct) but it will also fail to find `Mr. Smith` from `<span class="title">Mr.</span> Smith`.

This is relevant for the first case too:

> Inserting text before/after a string.

Because it may not be obvious if a chunk of string reported by `get_modifiable_text()` is connected or not to the one before or after it.

---

I do advise caution about widely teaching `set_modifiable_text()` as _the way to modify text_ because it's still a relatively internal method. It's useful to low-level utilities, but ultimately `set_inner_text()` and `set_inner_html()` will be the hallmark public methods.

---

Something like replacing Emoji seems like one of the more fitting examples to demonstrate, or replacing usernames or some kind of DSL embedded within the HTML.

**Reply by justintadlock** [reply]
Thanks, @dmsnell. That's just the sort of feedback I was thinking about.

**Comment by flexseth** [comment]
I have a great idea for this feature that will be my first plugin submitted to the WP.org plugins repo!

Following the discussion for progress updates.

**Comment by ndiego** [comment]
Even if there are only a few examples, it might be good to have an article that covers the dos and don'ts of `set_modifiable_text()`.

**Comment by flexseth** [comment]
Is there anything that will need to be escaped when using the hook? Might be worth a mention

**Reply by justintadlock** [reply]
I believe plain text is run through `htmlspecialchars()` with the method, but I haven't really dove into the code.

**Reply by dmsnell** [reply]
there should be no escaping necessary for _any_ methods in the HTML API, including `set_modifiable_text()`. If you find any cases where it would be necessary, we consider that a bug and want to fix it.

**Comment by justintadlock** [comment]
I'm leaning toward closing this one in favor of https://github.com/WordPress/developer-blog-content/discussions/313

Let's get some of those foundational examples in place and wait for `set_inner_html()` for some of the ideas around this.
